test/test_model.py

```diff
@@ -270,6 +270,22 @@ def test_update_app_relation_data(
             rel.data[harness.model.app].update(*args, **kwargs)
             assert harness.get_relation_data(relation_id, harness.model.app) == {'foo': 'baz'}
 
+    def test_set_app_relation_data(self, harness: ops.testing.Harness[ops.CharmBase]):
+        relation_name = 'db1'
+        original_data = {'hello': 'world'}
+        new_data = {'goodnight': 'moon'}
+        relation_id = harness.add_relation(relation_name, 'remote', app_data=original_data)
+        relation = harness.model.get_relation(relation_name)
+        assert relation is not None
+        assert relation.data[relation.app] == original_data
+        # new relation data can be set via RelationData.__setitem__
+        relation.data[relation.app] = new_data
+        assert relation.data[relation.app] == new_data
+        assert harness.get_relation_data(relation_id, relation.app) == new_data
```

Comment:
Do we also want to assert that the `RelationDataContent` object (id) hasn't changed?